File: src/main/codebase/projectJargon.ts
import * as fs from 'fs';
import * as path from 'path';

/**
 * Deterministic, LLM-free project-jargon extraction.
 *
 * The richer {@link CodebaseAnalyzer} scan asks an LLM to write a terminology
 * document, which is great but requires the `claude` CLI, a network round-trip,
 * and an explicit user action. This module instead mines identifiers directly
 * from a repo — dependency names from manifests, exported symbol names, and
 * source file names — so jargon biasing works with **zero setup** and even when
 * `llmProvider` is `none`. The result feeds {@link buildSpeechBiasPrompt}.
 */

const SKIP_DIRS = new Set([
  'node_modules', '.git', 'target', 'dist', 'build', '.next', 'out', 'vendor',
  '.venv', 'venv', '__pycache__', 'coverage', '.turbo', 'Pods', 'DerivedData',
  '.cache', '.idea', '.vscode',
]);

const SOURCE_EXT = new Set([
  '.ts', '.tsx', '.js', '.jsx', '.mjs', '.cjs', '.rs', '.py', '.go', '.java',
  '.kt', '.swift', '.rb', '.php', '.cs', '.c', '.cc', '.cpp', '.h', '.hpp',
  '.scala', '.ex', '.exs', '.dart',
]);

// Declarations whose name is a term a developer would say out loud. Language
// agnostic on purpose — one sweep covers JS/TS, Rust, Python, Go, Swift, etc.
const SYMBOL_PATTERNS: RegExp[] = [
  /\bexport\s+(?:default\s+)?(?:async\s+)?function\s+([A-Za-z_$][\w$]+)/g,
  /\bexport\s+(?:const|let|var|class|type|interface|enum|abstract\s+class)\s+([A-Za-z_$][\w$]+)/g,
  /\b(?:class|interface|type|enum)\s+([A-Za-z_$][\w$]+)/g,
  /\bdef\s+([A-Za-z_][\w]+)/g,                              // python
  /\bfunc\s+([A-Za-z_][\w]+)/g,                             // go / swift
  /\b(?:pub\s+)?(?:async\s+)?fn\s+([A-Za-z_][\w]+)/g,       // rust
  /\b(?:pub\s+)?(?:struct|enum|trait|mod)\s+([A-Za-z_][\w]+)/g, // rust
];

export interface JargonOptions {
  /** Stop after scanning this many source files. */
  maxFiles?: number;
  /** Return at most this many terms (dependency + symbol + filename, in priority order). */
  maxTerms?: number;
  /** Read at most this many bytes per source file. */
  maxBytesPerFile?: number;
}

/** True for tokens worth biasing on: multi-char and identifier-shaped. */
function isUsefulTerm(tok: string): boolean {
  return tok.length >= 3 && tok.length <= 40 && /^[A-Za-z][A-Za-z0-9_.-]*$/.test(tok);
}

/** True for a filename base that reads like a symbol (not "index"/"main"/"utils"). */
function looksTechnical(tok: string): boolean {
  return /[a-z][A-Z]/.test(tok) || tok.includes('_') || tok.includes('-') || /^[A-Z]{2,}/.test(tok);
}

function readManifestDeps(projectPath: string): string[] {
  const out: string[] = [];
  const tryRead = (file: string): string | null => {
    try { return fs.readFileSync(path.join(projectPath, file), 'utf-8'); } catch { return null; }
  };

  // package.json — name + all dependency keys.
  const pkg = tryRead('package.json');
  if (pkg) {
    try {
      const json = JSON.parse(pkg);
      if (typeof json.name === 'string') out.push(json.name.replace(/^@[^/]+\//, ''));
      for (const field of ['dependencies', 'devDependencies', 'peerDependencies', 'optionalDependencies']) {
        if (json[field] && typeof json[field] === 'object') {
          for (const dep of Object.keys(json[field])) out.push(dep.replace(/^@[^/]+\//, ''));
        }
      }
    } catch { /* malformed package.json — skip */ }
  }

  // Cargo.toml — package name + dependency keys (simple section parse).
  const cargo = tryRead('Cargo.toml');
  if (cargo) {
    let section = '';
    for (const raw of cargo.split('\n')) {
      const line = raw.trim();
      const sec = line.match(/^\[([^\]]+)\]/);
      if (sec) { section = sec[1]; continue; }
      const nameMatch = line.match(/^name\s*=\s*"([^"]+)"/);
      if (section === 'package' && nameMatch) { out.push(nameMatch[1]); continue; }
      if (/dependencies$/.test(section)) {
        const dep = line.match(/^([A-Za-z0-9_-]+)\s*=/);
        if (dep) out.push(dep[1]);
      }
    }
  }

  // requirements.txt — python package names before any version specifier.
  const reqs = tryRead('requirements.txt');
  if (reqs) {
    for (const raw of reqs.split('\n')) {
      const m = raw.trim().match(/^([A-Za-z0-9_.-]+)/);
      if (m && !raw.trim().startsWith('#')) out.push(m[1]);
    }
  }

  // go.mod — module name + required module last path segments.
  const gomod = tryRead('go.mod');
  if (gomod) {
    for (const raw of gomod.split('\n')) {
      const m = raw.trim().match(/(?:module|require)?\s*[\w./-]*?([A-Za-z0-9_-]+)(?:\s+v[\d.]+)?\s*$/);
      if (m && m[1].length >= 3) out.push(m[1]);
    }
  }

  return out;
}

function collectSymbolsAndNames(projectPath: string, maxFiles: number, maxBytesPerFile: number): string[] {
  const out: string[] = [];
  const stack: string[] = [projectPath];
  let scanned = 0;

  while (stack.length && scanned < maxFiles) {
    const dir = stack.pop()!;
    let entries: fs.Dirent[];
    try { entries = fs.readdirSync(dir, { withFileTypes: true }); } catch { continue; }

    for (const entry of entries) {
      if (scanned >= maxFiles) break;
      const name = entry.name;
      if (name.startsWith('.') && name !== '.') continue;
      const full = path.join(dir, name);

      if (entry.isDirectory()) {
        if (!SKIP_DIRS.has(name)) stack.push(full);
        continue;
      }
      if (!entry.isFile()) continue;

      const ext = path.extname(name);
      if (!SOURCE_EXT.has(ext)) continue;
      scanned++;

      const base = path.basename(name, ext);
      if (looksTechnical(base)) out.push(base);

      let content: string;
      try {
        const fd = fs.openSync(full, 'r');
        const buf = Buffer.alloc(maxBytesPerFile);
        const bytes = fs.readSync(fd, buf, 0, maxBytesPerFile, 0);
        fs.closeSync(fd);
        content = buf.toString('utf-8', 0, bytes);
      } catch { continue; }

      for (const pattern of SYMBOL_PATTERNS) {
        pattern.lastIndex = 0;
        for (const m of content.matchAll(pattern)) {
          if (m[1]) out.push(m[1]);
        }
      }
    }
  }

  return out;
}

/**
 * Extract jargon terms from a project directory without any LLM call.
 * Priority order: manifest dependency names → exported symbols/filenames.
 */
export function extractProjectJargon(projectPath: string, opts: JargonOptions = {}): string[] {
  const maxFiles = opts.maxFiles ?? 300;
  const maxTerms = opts.maxTerms ?? 400;
  const maxBytesPerFile = opts.maxBytesPerFile ?? 200_000;

  let resolved = projectPath;
  if (resolved.startsWith('~')) resolved = resolved.replace(/^~/, process.env.HOME || '');

  const ordered = [
    ...readManifestDeps(resolved),
    ...collectSymbolsAndNames(resolved, maxFiles, maxBytesPerFile),
  ];

  const seen = new Set<string>();
  const terms: string[] = [];
  for (const raw of ordered) {
    const tok = raw.trim();
    if (!isUsefulTerm(tok)) continue;
    const key = tok.toLowerCase();
    if (seen.has(key)) continue;
    seen.add(key);
    terms.push(tok);
    if (terms.length >= maxTerms) break;
  }
  return terms;
}

File: src-tauri/src/codebase/project_jargon.rs
//! Deterministic, LLM-free project-jargon extraction.
//!
//! Mirrors `src/main/codebase/projectJargon.ts`. Mines identifiers directly from
//! a repo — dependency names from manifests, exported symbol names, and source
//! file names — so STT biasing works with zero setup and no LLM configured.

use std::collections::HashSet;
use std::fs;
use std::io::Read;
use std::path::{Path, PathBuf};

use regex::Regex;

const SKIP_DIRS: &[&str] = &[
    "node_modules", ".git", "target", "dist", "build", ".next", "out", "vendor",
    ".venv", "venv", "__pycache__", "coverage", ".turbo", "Pods", "DerivedData",
    ".cache", ".idea", ".vscode",
];

const SOURCE_EXT: &[&str] = &[
    "ts", "tsx", "js", "jsx", "mjs", "cjs", "rs", "py", "go", "java", "kt",
    "swift", "rb", "php", "cs", "c", "cc", "cpp", "h", "hpp", "scala", "ex", "exs", "dart",
];

fn symbol_patterns() -> Vec<Regex> {
    vec![
        Regex::new(r"\bexport\s+(?:default\s+)?(?:async\s+)?function\s+([A-Za-z_$][\w$]+)").unwrap(),
        Regex::new(r"\bexport\s+(?:const|let|var|class|type|interface|enum|abstract\s+class)\s+([A-Za-z_$][\w$]+)").unwrap(),
        Regex::new(r"\b(?:class|interface|type|enum)\s+([A-Za-z_$][\w$]+)").unwrap(),
        Regex::new(r"\bdef\s+([A-Za-z_][\w]+)").unwrap(),
        Regex::new(r"\bfunc\s+([A-Za-z_][\w]+)").unwrap(),
        Regex::new(r"\b(?:pub\s+)?(?:async\s+)?fn\s+([A-Za-z_][\w]+)").unwrap(),
        Regex::new(r"\b(?:pub\s+)?(?:struct|enum|trait|mod)\s+([A-Za-z_][\w]+)").unwrap(),
    ]
}

fn is_useful_term(tok: &str) -> bool {
    let len = tok.chars().count();
    if len < 3 || len > 40 {
        return false;
    }
    let re = Regex::new(r"^[A-Za-z][A-Za-z0-9_.-]*$").unwrap();
    re.is_match(tok)
}

fn looks_technical(tok: &str) -> bool {
    let camel = Regex::new(r"[a-z][A-Z]").unwrap();
    let upper_start = Regex::new(r"^[A-Z]{2,}").unwrap();
    camel.is_match(tok) || tok.contains('_') || tok.contains('-') || upper_start.is_match(tok)
}

fn strip_scope(name: &str) -> String {
    // "@scope/pkg" -> "pkg"
    match name.rfind('/') {
        Some(idx) if name.starts_with('@') => name[idx + 1..].to_string(),
        _ => name.to_string(),
    }
}

fn read_manifest_deps(project_path: &Path) -> Vec<String> {
    let mut out: Vec<String> = Vec::new();
    let read = |file: &str| -> Option<String> { fs::read_to_string(project_path.join(file)).ok() };

    // package.json — name + all dependency keys.
    if let Some(pkg) = read("package.json") {
        if let Ok(json) = serde_json::from_str::<serde_json::Value>(&pkg) {
            if let Some(name) = json.get("name").and_then(|v| v.as_str()) {
                out.push(strip_scope(name));
            }
            for field in ["dependencies", "devDependencies", "peerDependencies", "optionalDependencies"] {
                if let Some(map) = json.get(field).and_then(|v| v.as_object()) {
                    for key in map.keys() {
                        out.push(strip_scope(key));
                    }
                }
            }
        }
    }

    // Cargo.toml — package name + dependency keys.
    if let Some(cargo) = read("Cargo.toml") {
        let mut section = String::new();
        let name_re = Regex::new(r#"^name\s*=\s*"([^"]+)""#).unwrap();
        let dep_re = Regex::new(r"^([A-Za-z0-9_-]+)\s*=").unwrap();
        for raw in cargo.lines() {
            let line = raw.trim();
            if let Some(sec) = line.strip_prefix('[').and_then(|s| s.strip_suffix(']')) {
                section = sec.to_string();
                continue;
            }
            if section == "package" {
                if let Some(c) = name_re.captures(line) {
                    out.push(c[1].to_string());
                    continue;
                }
            }
            if section.ends_with("dependencies") {
                if let Some(c) = dep_re.captures(line) {
                    out.push(c[1].to_string());
                }
            }
        }
    }

    // requirements.txt — python package names before any version specifier.
    if let Some(reqs) = read("requirements.txt") {
        let re = Regex::new(r"^([A-Za-z0-9_.-]+)").unwrap();
        for raw in reqs.lines() {
            let line = raw.trim();
            if line.starts_with('#') {
                continue;
            }
            if let Some(c) = re.captures(line) {
                out.push(c[1].to_string());
            }
        }
    }

    out
}

fn collect_symbols_and_names(
    project_path: &Path,
    max_files: usize,
    max_bytes_per_file: usize,
) -> Vec<String> {
    let mut out: Vec<String> = Vec::new();
    let mut stack: Vec<PathBuf> = vec![project_path.to_path_buf()];
    let mut scanned = 0usize;
    let patterns = symbol_patterns();

    while let Some(dir) = stack.pop() {
        if scanned >= max_files {
            break;
        }
        let entries = match fs::read_dir(&dir) {
            Ok(e) => e,
            Err(_) => continue,
        };
        for entry in entries.flatten() {
            if scanned >= max_files {
                break;
            }
            let name = entry.file_name().to_string_lossy().to_string();
            if name.starts_with('.') {
                continue;
            }
            let full = entry.path();
            let file_type = match entry.file_type() {
                Ok(t) => t,
                Err(_) => continue,
            };
            if file_type.is_dir() {
                if !SKIP_DIRS.contains(&name.as_str()) {
                    stack.push(full);
                }
                continue;
            }
            if !file_type.is_file() {
                continue;
            }
            let ext = full.extension().and_then(|e| e.to_str()).unwrap_or("");
            if !SOURCE_EXT.contains(&ext) {
                continue;
            }
            scanned += 1;

            if let Some(base) = full.file_stem().and_then(|b| b.to_str()) {
                if looks_technical(base) {
                    out.push(base.to_string());
                }
            }

            // Read a bounded prefix of the file.
            let mut content = String::new();
            if let Ok(mut f) = fs::File::open(&full) {
                let mut buf = vec![0u8; max_bytes_per_file];
                if let Ok(n) = f.read(&mut buf) {
                    content = String::from_utf8_lossy(&buf[..n]).to_string();
                }
            }
            for pattern in &patterns {
                for cap in pattern.captures_iter(&content) {
                    if let Some(m) = cap.get(1) {
                        out.push(m.as_str().to_string());
                    }
                }
            }
        }
    }

    out
}

/// Extract jargon terms from a project directory without any LLM call.
/// Priority order: manifest dependency names → exported symbols/filenames.
pub fn extract_project_jargon(project_path: &str) -> Vec<String> {
    let max_files = 300usize;
    let max_terms = 400usize;
    let max_bytes_per_file = 200_000usize;

    let resolved = if project_path.starts_with('~') {
        project_path.replacen('~', &dirs::home_dir().unwrap_or_default().to_string_lossy(), 1)
    } else {
        project_path.to_string()
    };
    let root = Path::new(&resolved);

    let mut ordered = read_manifest_deps(root);
    ordered.extend(collect_symbols_and_names(root, max_files, max_bytes_per_file));

    let mut seen: HashSet<String> = HashSet::new();
    let mut terms: Vec<String> = Vec::new();
    for raw in ordered {
        let tok = raw.trim();
        if !is_useful_term(tok) {
            continue;
        }
        let key = tok.to_lowercase();
        if seen.insert(key) {
            terms.push(tok.to_string());
            if terms.len() >= max_terms {
                break;
            }
        }
    }
    terms
}

fn jargon_path() -> PathBuf {
    dirs::home_dir().unwrap_or_default()
        .join("Library/Application Support/echo/project-jargon.json")
}

/// Deterministic jargon scan + cache to disk for STT biasing.
pub fn quick_scan(project_path: &str) -> Vec<String> {
    let terms = extract_project_jargon(project_path);
    let path = jargon_path();
    if let Some(dir) = path.parent() {
        fs::create_dir_all(dir).ok();
    }
    if let Ok(json) = serde_json::to_string(&terms) {
        fs::write(&path, json).ok();
    }
    terms
}

/// Load cached deterministic jargon terms (empty if none).
pub fn load_jargon_terms() -> Vec<String> {
    let path = jargon_path();
    if let Ok(content) = fs::read_to_string(&path) {
        if let Ok(terms) = serde_json::from_str::<Vec<String>>(&content) {
            return terms;
        }
    }
    Vec::new()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn keeps_only_useful_terms() {
        assert!(is_useful_term("runPipeline"));
        assert!(is_useful_term("api_key"));
        assert!(!is_useful_term("ab"));
        assert!(!is_useful_term("no spaces here"));
    }

    #[test]
    fn technical_filename_detection() {
        assert!(looks_technical("speechBias"));
        assert!(looks_technical("app_profiles"));
        assert!(looks_technical("API"));
        assert!(!looks_technical("index"));
        assert!(!looks_technical("main"));
    }

    #[test]
    fn strips_npm_scope() {
        assert_eq!(strip_scope("@tauri-apps/api"), "api");
        assert_eq!(strip_scope("react"), "react");
    }
}
